tst/CTA.Rules.Test/AwsRulesBaseTest.cs

```diff
@@ -253,7 +254,39 @@ protected string CopySolutionFolderToTemp(string solutionName, string tempDir)
             string solutionPath = Directory.EnumerateFiles(tempDir, solutionName, SearchOption.AllDirectories).FirstOrDefault(s => !s.Contains(string.Concat(Path.DirectorySeparatorChar, CopyFolder, Path.DirectorySeparatorChar)));
             string solutionDir = Directory.GetParent(solutionPath).FullName;
             var newTempDir = Path.Combine(GetTstPath(this.GetType()), CopyFolder, Guid.NewGuid().ToString());
-            CopyDirectory(new DirectoryInfo(solutionDir), new DirectoryInfo(newTempDir));
+
+            int folderCount = 1;
+
+            if (solutionPath.Contains(".sln") && File.Exists(solutionPath))
+            {
+                IEnumerable<string> projects = Utils.GetProjectPaths(solutionPath);
+
+                int depths = projects.ToList().Max(p => Regex.Matches(Path.GetRelativePath(solutionDir, p), Regex.Escape("..")).Count);
+                for (int i = 0; i < depths; i++)
+                {
+                    newTempDir += "\\Folder" + folderCount++;
+                }
+
+                CopyDirectory(new DirectoryInfo(solutionDir), new DirectoryInfo(newTempDir));
+
+                foreach (string project in projects)
+                {
+                    string projPath = Directory.GetParent(project).FullName;
+
+                    if (!Utils.IsSubPathOf(solutionDir, projPath))
+                    {
+                        string relativeSrc = Path.GetRelativePath(solutionDir, projPath);
+                        string projName = Path.GetFileName(project);
+                        string newRelDir = Path.Combine(newTempDir, relativeSrc);
+
+                        Utils.CopyFolderToTemp(projName, projPath, newRelDir);
+                    }
+                }
+            }
```

Comment:
Correct me if I'm wrong, but my understanding here is that if what we're copying is not a solution folder (i.e. if it's a solution file), then do something different (i.e. build a folder structure that supports copying any projects that don't lie within the solution directory's direct path).

If that's the case, can we put this logic in a separate function called `CopySolutionAndProjects` or something similar? This will reduce complexity of the existing method and prevent misunderstanding of what this method does (the name assumes the input will be a directory). We may also want to add helper functions for the other stuff that's going on, for example the part where we rebuild the relative file structure for the project files.

Reply:
I will extract the copy external projects portion to another method as you suggested.

Reply:
Will also extract the rebuilding of the relative file structure, thanks good ideas :-)